### PR TITLE
[api-extractor] Issue a warning for forgotten exports

### PR DESCRIPTION
**Pull request by octogonz**
API Extractor 6 issued a warning if an exported symbol was found that referenced another local symbol that was not exported:

```ts
import { IWidget } from 'widget';

interface IOptions { // <-- missing "export" keyword
  x: number;
}

interface IEvent {
  sender: string;
}

class Base { // <-- missing "export" keyword
}

// "Base" is a forgotten export because it is part of the definition of Child.
export class Child extends Base {
  // "IOptions" is a forgotten export.  Callers will not be able to refer to the
  // interface by name.
  constructor(options: IOptions) {
  }

  // "IWidget" is NOT a forgotten export, because it is not a local symbol.
  // Callers can import it from the "widget" package if they like.
  public widget: IWidget;

  // IEvent is NOT a forgotten export, because it is a private implementation detail.
  private _event: IEvent;
}
```

This PR reintroduces this validation for the API Extractor 7 analyzer.

### CODE REVIEW

File: apps/api-extractor/src/api/Extractor.ts [hunk 1]
```diff
@@ -391,6 +392,8 @@ export class Extractor {
 
     collector.analyze();
 
+    VisibilityChecker.check(collector);
```
Comment (iclanton) [inline apps/api-extractor/src/api/Extractor.ts]: This pattern is odd to me. It's like this is a plugin, but its applied by the the application directly.

File: apps/api-extractor/src/collector/VisibilityChecker.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
+import * as path from 'path';
+import * as ts from 'typescript';
+
+import { Collector } from './Collector';
+import { AstSymbol } from '../analyzer/AstSymbol';
+import { AstDeclaration } from '../analyzer/AstDeclaration';
+// import { SymbolMetadata } from './SymbolMetadata';
```
Comment (iclanton) [inline apps/api-extractor/src/collector/VisibilityChecker.ts]: Remove?

File: apps/api-extractor/src/collector/VisibilityChecker.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        // whether it's exported.
+        //
+        // TODO: Technically we should also check each of the nested scopes along the way.
+        const rootSymbol: AstSymbol = referencedEntity.rootAstSymbol;
+
+        if (!rootSymbol.isExternal) {
+          const collectorEntity: CollectorEntity | undefined = collector.tryGetCollectorEntity(rootSymbol);
+
+          if (collectorEntity && collectorEntity.exported) {
+            // const metadata: SymbolMetadata = collector.fetchMetadata(referencedEntity);
```
Comment (iclanton) [inline apps/api-extractor/src/collector/VisibilityChecker.ts]: ?

File: apps/api-extractor/src/collector/VisibilityChecker.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+              // The main usage scenario for ECMAScript symbols is to attach private data to a JavaScript object,
+              // so as a special case, we do NOT report them as forgotten exports.
+              if (!VisibilityChecker._isEcmaScriptSymbol(referencedEntity)) {
+
+                collector.messageRouter.addAnalyzerIssue(ExtractorMessageId.ForgottenExport,
+                  `The symbol "${rootSymbol.localName}" needs to be exported`
+                    + ` by the entry point ${entryPointFilename}`,
+                  astDeclaration);
+              }
+
```
Comment (iclanton) [inline apps/api-extractor/src/collector/VisibilityChecker.ts]: There is weird whitespace in this file. Can you remove the random newlines?

File: apps/api-extractor/src/collector/VisibilityChecker.ts [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+            }
+
+          }
+        }
+      }
+    }
+  }
+
+  // Detect an AstSymbol that refers to an ECMAScript symbol declaration such as:
```
Comment (iclanton) [inline apps/api-extractor/src/collector/VisibilityChecker.ts]: Should this be a TSDoc comment?

File: apps/api-extractor/src/collector/VisibilityChecker.ts [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+  //
+  // const mySymbol: unique symbol = Symbol('mySymbol');
+  private static _isEcmaScriptSymbol(astSymbol: AstSymbol): boolean {
+    if (astSymbol.astDeclarations.length !== 1) {
+      return false;
+    }
+
+    // We are matching a form like this:
```
Comment (iclanton) [inline apps/api-extractor/src/collector/VisibilityChecker.ts]: Same with this one.

File: build-tests/api-extractor-scenarios/etc/test-outputs/ambientNameConflict/api-extractor-scenarios.api.ts [hunk 7]
```diff
@@ -1,3 +1,5 @@
+// Warning: (ae-forgotten-export) The symbol "Promise" needs to be exported by the entry point index.d.ts
```
Comment (iclanton) [inline build-tests/api-extractor-scenarios/etc/test-outputs/ambientNameConflict/api-extractor-scenarios.api.ts]: Are these issues validated in any non-manual way? Maybe these files should be Jest snapshots instead.

Comment (iclanton) [reply build-tests/api-extractor-scenarios/etc/test-outputs/ambientNameConflict/api-extractor-scenarios.api.ts]: I know changes here create diffs, but I'm not sure it's obvious that these diffs are test outputs.

Maybe they are...

File: apps/api-extractor/src/api/Extractor.ts (hunk 1 above)
Comment (octogonz) [reply apps/api-extractor/src/api/Extractor.ts]: I wanted the visibility checking logic to be in its own source file, but currently it doesn't maintain any state.  That's why I used a static class.

File: apps/api-extractor/src/collector/VisibilityChecker.ts (hunk 2 above)
Comment (octogonz) [reply apps/api-extractor/src/collector/VisibilityChecker.ts]: I was starting to implement consistency checking (e.g. an `@beta` function should not return an `@alpha` type), but I think I'll do that as a separate PR.

File: build-tests/api-extractor-scenarios/etc/test-outputs/ambientNameConflict/api-extractor-scenarios.api.ts (hunk 7 above)
Comment (octogonz) [reply build-tests/api-extractor-scenarios/etc/test-outputs/ambientNameConflict/api-extractor-scenarios.api.ts]: We really need a build script that will break a production build if the contents of these file has changed, but the person forgot to commit those changes to Git.  Right now I simply have to remember to do it when I work on API Extractor.

Out of scope for this PR, though.

**Pull request merged**